### ISSUE TITLE
Rationalize parsers

### ISSUE THREAD

**Issue by jnfoster** [frenetic-lang/frenetic#339]
We currently have two parsers: NetKAT_Parser and the Camlp4 syntax extension. They are inconsistent which is confusing. Either we need to eliminate NetKAT_Parser and expose the equivapent functions from Camlp4 (my strong preference) or we need to add suitable test infrastructure and social processes so they are not able to diverge... 

(This tripped me up this morning...)

-N

**Comment by craig-riecke** [comment]
This is scheduled to be fixed as part of general cleanup in the Unified branch.  Unfortunately, it looks like we still have to do two parsers - a NetKat sytnax extension that spits out OCaml AST's and a plain ol' parser that outputs native NetKat.  If you try to load the syntax extension while compiling Frentic-proper, it thinks the $ are antiquotes.  So I gotta keep that separate.  

On the good side, the Lexer can be shared between them and they'll both be camlp4.  That's an improvement.

**Comment by craig-riecke** [comment]
Ugggh.  It looks like camlp4 and cmdliner dislike each other.  I pushed the camlp4 + ulex based lexer into the main Frenetic library and wrote a camlp4 parser that outputs pure NetKAT data structures.  That works ... or at least it works when writing tests against the library.  But when you try to build the executable frenetic.ml, the $ operator of cmdliner will not compile.  The error is:

```
+ /home/vagrant/.opam/system/bin/ocamlfind ocamldep -syntax camlp4o -syntax camlp4o -syntax camlp4o -syntax camlp4o -package yojson -package ulex -package tcpip -package str -package sexplib.syntax -package sexplib -package ocamlgraph -package fieldslib.syntax -package fieldslib -package cstruct.syntax -package cstruct.async -package cstruct -package core -package camlp4.lib -package camlp4.extend -package camlp4 -package base64 -package async -package cohttp.async -package cohttp -package cmdliner -package async_extended -modules frenetic/frenetic/frenetic.ml > frenetic/frenetic/frenetic.ml.depends
File "frenetic/frenetic/frenetic.ml", line 49, characters 42-51:
Parse error: [semi] expected after [str_item] (in [implem])
File "frenetic/frenetic/frenetic.ml", line 1:
Error: Error while running external preprocessor
Command line: camlp4 '-I' '/usr/lib/ocaml/camlp4' '-I' '/home/vagrant/.opam/system/lib/ulex' '-I' '/home/vagrant/.opam/system/lib/type_conv' '-I' '/usr/lib/ocaml' '-I' '/usr/lib/ocaml' '-I' '/home/vagrant/.opam/system/lib/sexplib' '-I' '/home/vagrant/.opam/system/lib/sexplib' '-I' '/home/vagrant/.opam/system/lib/fieldslib' '-I' '/home/vagrant/.opam/system/lib/fieldslib' '-I' '/home/vagrant/.opam/system/lib/cstruct' '-I' '/usr/lib/ocaml/camlp4' '-parser' 'o' '-parser' 'op' '-printer' 'p' 'pa_ulex.cma' 'pa_type_conv.cma' 'unix.cma' 'bigarray.cma' 'sexplib.cma' 'pa_sexp_conv.cma' 'fieldslib.cma' 'pa_fields_conv.cma' 'cstruct-syntax.cma' '-parser' 'Camlp4GrammarParser'  'frenetic/frenetic/frenetic.ml' > /tmp/ocamlppd87a08
```

However, it looks like $ is a synonym for the function "app" in CmdLiner, so I'm gonna see if I can replace it...

**Comment by arjunguha** [comment]
yuck, that's terrible.

**Comment by craig-riecke** [comment]
Yeah ... even worse, camlp4 tries to parse $ in comments, so you can't even comment out the code properly!  But replacing the $ operator with the straight function call Term.app worked ... it's not as pretty, but functional.

So the latest commit to unified branch replaces the old parser with the new Camlp4 one.  There were some discrepancies between the two - for example, "&&" was used instead of "and" and the "|" was not recognized.  But after squaring them up, the unit tests that exercised the old parser work on the new one as well, so I think we're set.

I will close this once Unified is merged into Master.  That might be as early as tomorrow, 5/20.

**Comment by craig-riecke** [comment]
I guess the Ocaml manual has the last word on this subject (http://caml.inria.fr/pub/docs/manual-ocaml/lex.html):

Note that the following identifiers are keywords of the Camlp4 extensions and should be avoided for compatibility reasons.

    parser    value    $     $$    $:    <:    <<    >>    ??

**Comment by jnfoster** [comment]
We should open an issue in Cmdliner.

-N

**Comment by craig-riecke** [comment]
Done!  https://github.com/dbuenzli/cmdliner/issues/32
